<h2>Why is this an issue?</h2>
<p>This rule applies whenever an <code>if</code> statement is followed by one or more <code>elsif</code> statements; the final <code>elsif</code>
should be followed by an <code>else</code> statement.</p>
<p>The requirement for a final <code>else</code> statement is defensive programming.</p>
<p>The <code>else</code> statement should either take appropriate action or contain a suitable comment as to why no action is taken. This is
consistent with the requirement to have a final <code>else</code> clause in a <code>case</code> statement.</p>
<h3>Noncompliant code example</h3>
<pre>
if x == 0
  doSomething
elsif x == 1
  doSomethingElse
end
</pre>
<h3>Compliant solution</h3>
<pre>
if x == 0
  doSomething
elsif x == 1
  doSomethingElse
else
  raise 'An error has occured'
end
</pre>
<h3>Exceptions</h3>
<p>When all branches of an <code>if</code>-<code>else if</code> end with <code>return</code>, <code>break</code> or <code>throw</code>, the code that
comes after the <code>if</code> implicitly behaves as if it was in an <code>else</code> clause. This rule will therefore ignore that case.</p>

